File: JavaScript/7js 3/part.js
/*
|--------------------------------------------------------------------------
| Alan Partridge I - Partridge Watch
|--------------------------------------------------------------------------
|
*/

function part(arr) {
   let count = 0;
   let x = ["Partridge", "PearTree", "Chat", "Dan", "Toblerone", "Lynn", "AlphaPapa", "Nomad"];
   arr.forEach((i) => {
      if (x.includes(i)) count += 1;
   });
   return count != 0
      ? `Mine's a Pint${"!".repeat(count)}`
      : "Lynn, I've pierced my foot on a spike!!";
}


const part = (x) =>
   ((m) => (m ? "Mine's a Pint" + "!".repeat(m) : "Lynn, I've pierced my foot on a spike!!"))(
      x.reduce(
         (a, b) =>
            a +
            ([
               "Partridge",
               "PearTree",
               "Chat",
               "Dan",
               "Toblerone",
               "Lynn",
               "AlphaPapa",
               "Nomad",
            ].indexOf(b) != -1
               ? 1
               : 0),
         0
      )
   );


function part(x) {
   let list = ["Partridge", "PearTree", "Chat", "Dan", "Toblerone", "Lynn", "AlphaPapa", "Nomad"];
   let out = x.filter((el) => list.includes(el)) || [];
   return out.length
      ? "Mine's a Pint" + "!".repeat(out.length)
      : "Lynn, I've pierced my foot on a spike!!";
}


console.log(part(["Grouse", "Partridge", "Pheasant"]), "Mine's a Pint!");
console.log(
   part(["Pheasant", "Goose", "Starling", "Robin"]),
   "Lynn, I've pierced my foot on a spike!!"
);
console.log(
   part([
      "Grouse",
      "Partridge",
      "Pheasant",
      "Goose",
      "Starling",
      "Robin",
      "Thrush",
      "Emu",
      "PearTree",
      "Chat",
      "Dan",
      "Square",
      "Toblerone",
      "Lynn",
      "AlphaPapa",
      "BMW",
      "Graham",
      "Tool",
      "Nomad",
      "Finger",
      "Hamster",
   ]),
   "Mine's a Pint!!!!!!!!"
);

File: Python/7py 3/part.py
"""
|--------------------------------------------------------------------------
| Alan Partridge I - Partridge Watch
|--------------------------------------------------------------------------
|
"""

def part(arr):
   count, x = 0, ["Partridge", "PearTree", "Chat", "Dan", "Toblerone", "Lynn", "AlphaPapa", "Nomad"]
   for i in arr:
      if i in x: count += 1
   return f"Mine's a Pint{'!' * count}" if count != 0 else "Lynn, I've pierced my foot on a spike!!"


def part(arr):
   terms = {'Partridge', 'PearTree', 'Chat', 'Dan', 'Toblerone', 'Lynn', 'AlphaPapa', 'Nomad'}
   marks = ''.join('!' for x in arr if x in terms)
   return "Mine's a Pint" + marks if marks else "Lynn, I've pierced my foot on a spike!!"


def part(arr: list[str]) -> str:
   terms = ['Partridge', 'PearTree', 'Chat', 'Dan', 'Toblerone', 'Lynn', 'AlphaPapa', 'Nomad']
   count = 0
   for i in arr:        
      if i in terms: count += 1
   if count == 0: return "Lynn, I've pierced my foot on a spike!!"
   else: return "Mine's a Pint"+'!'*count


print(part(["Grouse", "Partridge", "Pheasant"]), "Mine's a Pint!")
print(part(["Pheasant", "Goose", "Starling", "Robin"]), "Lynn, I've pierced my foot on a spike!!")
print(part(["Grouse", "Partridge", "Partridge", "Partridge", "Pheasant"]), "Mine's a Pint!!!")
print(part([]), "Lynn, I've pierced my foot on a spike!!")
print(part(["Grouse", "Partridge", "Pheasant", "Goose", "Starling", "Robin", "Thrush", "Emu", "PearTree", "Chat", "Dan", "Square", "Toblerone", "Lynn", "AlphaPapa", "BMW", "Graham", "Tool", "Nomad", "Finger", "Hamster"]), "Mine's a Pint!!!!!!!!")